common/include/scipp/common/index_composition.h

```diff
@@ -0,0 +1,130 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2021 Scipp contributors (https://github.com/scipp)
+/// @file
+/// @author Jan-Lukas Wynen
+#pragma once
+
+#include <array>
+#include <cassert>
+#include <cstddef>
+#include <utility>
+
+#include "scipp/common/index.h"
+
+namespace scipp {
+/// Compute a flat index from strides and a multi-dimensional index.
+///
+/// @return sum_{i=0}^{ndim} ( strides[i] * indices[i] )
+/// @note This function uses *strides* and not a shape, meaning strides[d]
+///       is not the extent of the array in dimension d but rather the step
+///       length in the flat index to advance one element in d.
+///       It is thus *not* complementary to extract_indices.
+template <class ForwardIt1, class ForwardIt2>
+constexpr auto flat_index_from_strides(ForwardIt1 strides_it,
+                                       const ForwardIt2 strides_end,
+                                       ForwardIt2 indices_it) noexcept {
+  std::remove_const_t<std::remove_reference_t<decltype(*strides_it)>> result =
+      0;
+  for (; strides_it != strides_end; ++strides_it, ++indices_it) {
+    result += *strides_it * *indices_it;
+  }
+  return result;
+}
+
+/// Compute the bounds of a piece of memory.
+///
+/// Given a pointer `p` to some memory, a shape, and strides, this function
+/// returns a begin and an end index such that
+/// - `p + begin` is the smallest reachable address and
+/// - `p + end` is one past the largest reachable address
+///
+/// \return A pair of indices `{begin, end}`.
+template <class ForwardIt1, class ForwardIt2>
+constexpr auto memory_bounds(ForwardIt1 shape_it, const ForwardIt1 shape_end,
+                             ForwardIt2 strides_it) noexcept {
+  scipp::index begin = 0;
+  scipp::index end = 0;
+  for (; shape_it != shape_end; ++shape_it, ++strides_it) {
+    if (*strides_it < 0)
+      begin += *shape_it * *strides_it;
+    else
+      end += *shape_it * *strides_it;
+  }
+  // For scalars the above produces begin==end because all strides == 0.
```

Comment:
Scalar don't have strides==0, rather no strides? Is `begin == end` the correct check?

Reply:
True.

It is in principle possible to produce `begin == end` using negative strides. Is this something that can happen in practice?

Reply:
We should frequently get `begin == end` if a dimension has extent 0. Yes this happens in practice.

If strides are negative I would think we should *not* get `begin == end`, ever, unless extent is also 0?

Reply:
`begin` and `end` are cummulative over dimensions. If there is one dim with extent 0 but others with greater extend, we get `begin != end`. The only way the two can be equal is if either all dims have extent 0, we have a scalar, or if contriubtions from different dimensions cancel.

Reply:
Hmm, but for scalars we do *not* want `begin == end`? Shouldn't it be `end = begin + 1` in that case?

Reply:
I mean, that is what is being done here, but my points is there are legitimate cases with `begin == end`, but scalars are not, so only that case should be fixed here?

Reply:
Maybe just return early on top:
```cpp
if (shape_it == shape_end)  // scalar
  return {0, 1};
```